llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/IREEPyDM/IR/IREEPyDMOps.td

```diff
@@ -0,0 +1,521 @@
+    decay into.
+
+  }];
+
+  let arguments = (ins FlatSymbolRefAttr:$callee,
+                       Variadic<IREEPyDM_AnyValueType>:$operands);
+  let results = (outs
+      IREEPyDM_ExceptionResultType:$exc_result,
+      IREEPyDM_AnyValueType:$result);
+
+  let extraClassDeclaration = [{
+    StringRef getCallee() { return callee(); }
+    FunctionType getCalleeType();
+
+    /// Get the argument operands to the called function.
+    operand_range getArgOperands() {
+      return {arg_operand_begin(), arg_operand_end()};
+    }
+
+    operand_iterator arg_operand_begin() { return operand_begin(); }
+    operand_iterator arg_operand_end() { return operand_end(); }
+
+    /// Return the callee of this operation.
+    CallInterfaceCallable getCallableForCallee() {
+      return (*this)->getAttrOfType<SymbolRefAttr>("callee");
+    }
+  }];
+
+  let assemblyFormat = [{
+    $callee `(` $operands `)` attr-dict `:` functional-type($operands, results)
+  }];
+}
+
+def IREEPyDM_PatternMatchCallOp : IREEPyDM_Op<"pattern_match_call", [
+    DeclareOpInterfaceMethods<SymbolUserOpInterface>]> {
+  let summary = "Combines multiple functions together for generic dispatch";
+  let description = [{
+    This op enables specialized arity and type based dispatch via simple
+    pattern matching. It is generally used by the implementation to provide
+    both fully generic, runtime based implementations with the option to
+    select a specialization.
+
+    Alternatives are split into two groups:
+      - 'generic': Provides a maximally generic fallback implementation.
+        Generally an alternative will be matched purely based on arity and
+        structure of arguments. First match wins.
+      - 'specific': Provides a specific, strongly typed implementation. Matching
+        is by full type signature. First match wins.
+
+    Generally, during compilation, such calls will decay into regular calls
+    to one of the referenced, backing functions. How this is
+    done depends on phase:
+      - During canonicalization: A match to a specific function will be taken
+        immediately.
+      - After type inference: Any leftovers must be matched to a generic (by
+        an explicit pass) for runtime resolution.
+  }];
+
+  let arguments = (ins IREEPyDM_FlatSymbolRefArrayAttr:$generic_match,
+                       IREEPyDM_FlatSymbolRefArrayAttr:$specific_match,
+                       Variadic<IREEPyDM_AnyValueType>:$operands);
+  let results = (outs
+      IREEPyDM_ExceptionResultType:$exc_result,
+      IREEPyDM_AnyValueType:$result);
+
+  let assemblyFormat = [{
+    `(` $operands `)` `:` functional-type($operands, results)
+    `matching` `generic` $generic_match `specific` $specific_match
+    attr-dict
+  }];
+}
+
+def IREEPyDM_DynamicCallOp : IREEPyDM_PureOp<"dynamic_call", [
+    DeclareOpInterfaceMethods<SymbolUserOpInterface>]> {
+  let summary = "Dynamic dispatch to a `func` or `multi_func`.";
```

Comment:
do multi_funcs not exist yet or is this out of date?

Reply:
Old reference. I ended up backing multi_funcs out in favor of rolling the pattern matching into the PatternMatchCall op. Consider all of these to be placeholders that I'll fiddle with in transformations and find the right mix.